src/Processors/Transforms/DistinctTransform.cpp

```diff
@@ -48,6 +52,108 @@ void DistinctTransform::buildFilter(
     }
 }
 
+template <typename Method>
+void DistinctTransform::buildFilterWithMask(
+    Method & method,
+    const ColumnRawPtrs & columns,
+    IColumn::Filter & filter,
+    const size_t rows,
+    SetVariants & variants,
+    const IColumn::Filter & mask) const
+{
+    typename Method::State state(columns, key_sizes, nullptr);
+
+    for (size_t i = 0; i < rows; ++i)
+    {
+        if (!mask[i])
+        {
+            /// Already known duplicate row (by LC index), skip insertion
+            filter[i] = 0;
+            continue;
+        }
+
+        auto emplace_result = state.emplaceKey(method.data, i, variants.string_pool);
+        filter[i] = emplace_result.isInserted();
+    }
+}
+
+IColumn::Filter DistinctTransform::buildLowCardinalityMask(const ColumnLowCardinality & column, size_t num_rows)
+{
+    IColumn::Filter mask(num_rows, 0);
+
+    const auto & dictionary = column.getDictionary();
+    const auto dict_size = dictionary.size();
+
+    LCDictionaryKey dict_key;
+    dict_key.hash = dictionary.getHash();
+    dict_key.size = dict_size;
+
+    auto & state = lc_dict_states[dict_key];
+
+    /// The first time we see this dictionary, initialize the seen_indices array to keep track which entries
+    /// in the dictionary have been seen.
+    chassert(state.seen_count <= dict_size);
+    if (state.seen_indices.size() != dict_size)
+    {
+        chassert(state.seen_indices.empty());
+        chassert(state.seen_count == 0);
+        state.seen_indices.resize_fill(dict_size);
+    }
+
+    /// If we've already seen all dictionary indices for this dictionary,
+    /// then no row in this chunk (and also other chunks with the same dictionary) can produce a new distinct value.
+    if (state.seen_count == dict_size)
+        return mask;
+
+    auto & seen = state.seen_indices;
+
+    const auto index_type_size = column.getSizeOfIndexType();
+    const IColumn & indexes_column = *column.getIndexesPtr();
+
+    auto handle_index = [&](UInt64 idx, size_t row)
```

Comment:
Isn't `idx` a different type depending on the index_type? Seems like we are implicitly casting unnecessarily